modules/geo-layers/src/imagery-layer/imagery-layer.ts

```diff
@@ -0,0 +1,205 @@
+              layers: this.props.layers,
+              // todo image width may get out of sync with viewport width
+              width,
+              height,
+              bbox: bounds,
+              query_layers: this.props.layers,
+              x,
+              y,
+              info_format: 'text/plain'
+            })
+            // console.log(featureInfo);
+          }, 0);
+        }
+      }
+    });
+  }
+
+  /** Run a getMetadata on the image service */
+  async _loadMetadata(): Promise<void> {
+    this.props.onMetadataLoadStart();
+    const state = this.state as ImageryLayerState;
+    try {
+      state.metadata = await state.imageSource.getMetadata();
+      // technically we should get the latest layer after an async operation in case props have changed
+      // Although the response might no longer be expected
+      this.getCurrentLayer()?.props.onMetadataLoadComplete(state.metadata);
+    } catch (error) {
+      this.getCurrentLayer()?.props.onMetadataLoadError(error as Error);
+    }
+  }
+  
+  /** Load an image */
+  async loadImage(reason: string): Promise<void> {
+    const viewports = this.context.deck?.getViewports() || [];
+    if (viewports.length <= 0) {
+      return;
+    }
+
+    // TODO - need to handle multiple viewports
+    const viewport = viewports[0];
+    const bounds = viewport.getBounds();
+    const {width, height} = viewport;
+
+    const state = this.state as ImageryLayerState;
+
+    let requestId = getRequestId();
+
+    try {
+      this.props.onImageLoadStart(requestId);
+      const image = await state.imageSource.getImage({width, height, bbox: bounds, layers: this.props.layers});
+      Object.assign(this.state, {image, bounds, width, height});
+      this.getCurrentLayer()?.props.onImageLoadComplete(requestId);
+      this.setNeedsRedraw();
+    } catch (error) {
+      this.context.onError?.(error as Error, this);
+      this.getCurrentLayer()?.props.onImageLoadError(requestId, error as Error);
+    }    
+  }
+}
+
+/** Creates an image service if appropriate */
+function createImageSource(props: ImageryLayerProps): ImageSource {
+  if (typeof props.data === 'string') {
+    switch (props.serviceType) {
+      case 'template':
+        return new AdHocImageService({templateUrl: props.data});
+      case 'wms':
+      default: // currently only wms service supported
+        return new WMSService({serviceUrl: props.data})
+    }
+  }
+  if (props.data instanceof ImageSource) {
+    return props.data;
+  }
+  throw new Error('data props is not a valid image source');
+}
+
+// HELPERS
+
+let nextRequestId: number = 0;
+
+/** Global counter for issuing unique request ids */
+function getRequestId() {
+  return nextRequestId++;
+}
+
+let timeoutId;
+
+/** Runs an action in the future, cancels it if the new action is issued before it executes */
+function debounce(fn: Function, ms = 500): void {
+  clearTimeout(timeoutId);
+  timeoutId = setTimeout(() => fn(), ms);
+}
```

Comment:
These need to be wrapped in each ImageryLayer or you can't use multiple instances.

Reply:
debounce etc wrapped